tf2_geometry_msgs/include/tf2_geometry_msgs/tf2_geometry_msgs.hpp

```diff
@@ -347,6 +349,104 @@ template <>
 inline
   std::array<std::array<double, 6>, 6> getCovarianceMatrix(const geometry_msgs::msg::PoseWithCovarianceStamped& t)  {return covarianceRowMajorToNested(t.pose.covariance);}
 
+// Forward declaration
+void fromMsg(const geometry_msgs::msg::Transform& in, tf2::Transform& out);
+
+/** \brief Transform the covariance matrix of a PoseWithCovariance message to a new frame.
+ * \param cov_in The covariance matrix to transform.
+ * \param transform The transform to apply, as a Transform message.
+ * \return The transformed covariance matrix.
+ */
+inline
+geometry_msgs::msg::PoseWithCovariance::_covariance_type transformCovariance(const geometry_msgs::msg::PoseWithCovariance::_covariance_type& cov_in, const geometry_msgs::msg::Transform& transform)
+{
+    /**
+     * To transform a covariance matrix:
+     * 
+     * \verbatim[R 0] COVARIANCE [R' 0 ]
+      [0 R]            [0  R']\endverbatim
+     * 
+     * Where:
+     * 	R is the rotation matrix (3x3).
+     * 	R' is the transpose of the rotation matrix.
+     * 	COVARIANCE is the 6x6 covariance matrix to be transformed.
+     *
+     * Reference:
+     * 	A. L. Garcia, “Linear Transformations of Random Vectors,” in Probability,
+     * 	Statistics, and Random Processes For Electrical Engineering, 3rd ed.,
+     * 	Pearson Prentice Hall, 2008, pp. 320–322.
+     */
+
+    // get rotation matrix (and transpose)
+    tf2::Transform tf_transform;
+    fromMsg(transform, tf_transform);
+    const tf2::Matrix3x3 R = tf_transform.getBasis();
+
+    // convert covariance matrix into four 3x3 blocks
+    const tf2::Matrix3x3 cov_11(cov_in[0], cov_in[1], cov_in[2],
+			         cov_in[6], cov_in[7], cov_in[8],
+			         cov_in[12], cov_in[13], cov_in[14]);
+    const tf2::Matrix3x3 cov_12(cov_in[3], cov_in[4], cov_in[5],
+			         cov_in[9], cov_in[10], cov_in[11],
+			         cov_in[15], cov_in[16], cov_in[17]);
+    const tf2::Matrix3x3 cov_21(cov_in[18], cov_in[19], cov_in[20],
+			         cov_in[24], cov_in[25], cov_in[26],
+			         cov_in[30], cov_in[31], cov_in[32]);
+    const tf2::Matrix3x3 cov_22(cov_in[21], cov_in[22], cov_in[23],
+			         cov_in[27], cov_in[28], cov_in[29],
+			         cov_in[33], cov_in[34], cov_in[35]);
+
+    // perform blockwise matrix multiplication
+    const tf2::Matrix3x3 result_11 = R * cov_11 * R.transpose();
```

Comment:
`R.transpose()` returns [a new `Matrix3x3` instance](https://github.com/ros2/geometry2/blob/6fa2001c026ff104b890663b52f2e3d518140146/tf2/include/tf2/LinearMath/Matrix3x3.h#L557-L558), so it may be more efficient to store it in a local variable instead of calling `R.transpose()` multiple times.

This comment is premature optimization so feel free to ignore.

Reply:
Addressed in 8545842360dad058